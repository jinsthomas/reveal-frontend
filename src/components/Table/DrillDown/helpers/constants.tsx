/** Constants */
import React from 'react';

export const ID = 'id';
export type ID = typeof ID;

export const PARENT_ID = 'parent_id';
export type PARENT_ID = typeof PARENT_ID;

export const ROOT_PARENT_ID = null;
export type ROOT_PARENT_ID = typeof ROOT_PARENT_ID;

export const LINKER_ITEM_CSS_CLASS = 'dd-linker-item';
export type LINKER_ITEM_CSS_CLASS = typeof LINKER_ITEM_CSS_CLASS;

export const CLICKABLE_CSS_CLASS = 'dd-linker-item dd-clickable';
export type CLICKABLE_CSS_CLASS = typeof CLICKABLE_CSS_CLASS;

export const CARET_CSS_CLASS = 'dd-caret';
export type CARET_CSS_CLASS = typeof CARET_CSS_CLASS;

export const CARET_SPAN = <span className={CARET_CSS_CLASS}>&nbsp;&#9660;</span>;
export type CARET_SPAN = typeof CARET_SPAN;

export const DEFAULT_ROW_HEIGHT = '2em';
export type DEFAULT_ROW_HEIGHT = typeof DEFAULT_ROW_HEIGHT;
