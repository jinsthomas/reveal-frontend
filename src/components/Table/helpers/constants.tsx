/** Constants */
import React from 'react';
export const ID = 'id';
export type ID = typeof ID;

export const PARENT_ID = 'parent_id';
export type PARENT_ID = typeof PARENT_ID;

export const ROOT_PARENT_ID = null;
export type ROOT_PARENT_ID = typeof ROOT_PARENT_ID;

export const LINKER_ITEM_CSS_CLASS = 'dd-linker-item';
export type LINKER_ITEM_CSS_CLASS = typeof LINKER_ITEM_CSS_CLASS;

export const CLICKABLE_CSS_CLASS = 'dd-linker-item dd-clickable';
export type CLICKABLE_CSS_CLASS = typeof CLICKABLE_CSS_CLASS;

export const CARET_CSS_CLASS = 'dd-caret';
export type CARET_CSS_CLASS = typeof CARET_CSS_CLASS;

export const CARET = <span className={CARET_CSS_CLASS}>&nbsp;&#9660;</span>;
export type CARET = typeof CARET;

export const ROWS_TO_DISPLAY = 'Rows to display';
export type ROWS_TO_DISPLAY = typeof ROWS_TO_DISPLAY;

export const PAGE = 'page';
export type PAGE = typeof PAGE;

export const OF = 'of';
export type OF = typeof OF;
