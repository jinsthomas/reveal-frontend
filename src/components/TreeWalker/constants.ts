export const ACTIVE = 'Active';
export const FIND_BY_ID = 'location/findByJurisdictionIds';
export const FIND_BY_PROPERTIES = 'location/findByProperties';
export const LOCATION = 'location';
export const COULDNT_LOAD_PARENTS = 'Could not load parents';
