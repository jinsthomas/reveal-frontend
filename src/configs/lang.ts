// todo - init i18n here (or migrate everything below) and asynchronously export translations
// also - async export/imports article: https://medium.com/@WebReflection/javascript-dynamic-import-export-b0e8775a59d4

// Display Strings
export const WELCOME_TO_REVEAL = 'Welcome to Reveal';
export const GET_STARTED_MESSAGE = 'Get started by selecting an intervention below';
export const PRACTITIONER_CREATED_SUCCESSFULLY = 'Practitioner created Successfully.';
export const PRACTITIONER_EDITED_SUCCESSFULLY = 'Practitioner edited successfully.';
export const REMOVING_PRACTITIONER_FAILED = 'Removing Practitioner Failed';
export const ON_REROUTE_WITH_UNSAVED_CHANGES =
  'Unsaved Changes: please Save or Discard changes made';

export const HOME_TITLE = `Home page`;
export const IRS_TITLE = 'IRS';
export const IRS_REPORTING_TITLE = 'IRS Reporting';
export const CONDITIONAL_FORMATTING_RULES = 'Conditional formatting rules';
export const CONDITIONAL_FORMATTING_TITLE = 'Conditional Formatting';
export const FI_RESPONSE_ADHERENCE_TITLE = 'FI Response Adherence';
export const FOCUS_INVESTIGATION_START_TITLE = 'Focus Investigation Start';
export const DATE_CREATED = 'Date Created';

export const NEW_PLAN = 'New Plan';
export const ADD_PLAN = 'Add Plan';
export const CREATE_NEW_PLAN = 'Create New Plan';
export const UPDATE_PLAN = 'Update Plan';
export const CREATE_NEW_IRS_PLAN = 'Create New IRS Plan';
export const ADD_FOCUS_INVESTIGATION = 'Add Focus Investigation';

export const START_DATE = 'Start Date';
export const END_DATE = 'End Date';
export const NAME = 'Name';
export const LEGEND_LABEL = 'Legend';
export const DISCARD_CHANGES = 'Discard changes';
export const SAVE_CHANGES = 'Save changes';
export const ASSIGN_PLANS = 'Assign Plans';
export const ASSIGN_PRACTITIONERS = 'Assign Practitioners';

export const USERS = 'Users';
export const ABOUT = 'About';
export const LOGIN = 'Login';
export const SIGN_OUT = 'Sign Out';

export const MAP_LOAD_ERROR = 'Could not load the map';
export const AN_ERROR_OCCURED = 'An Error Ocurred';
export const PLEASE_FIX_THESE_ERRORS = 'Please fix these errors';
export const NO_INVESTIGATIONS_FOUND = 'No Investigations Found';
export const NO_PRACTITIONERS_ADDED_YET = 'No Practitioners Added yet';
export const PLAN_SELECT_PLACEHOLDER = 'Other plans in this area';
export const PLAN_STATUS_UPDATE_ERROR =
  'Sorry, something went wrong when we tried to update the plan status';
export const NO_PLAN_FOUND_ERROR = 'Sorry, no plan found in the cloud!';
export const SAVE_PLAN_DEFINITION_ERROR =
  'Uh oh, looks like something is (syntactically) wrong with the Plan schema';
export const SAVE_PLAN_NO_JURISDICTIONS_ERROR = 'Oops, no jurisdictions selected!';

export const SAVE_TEAM = 'Save Team';
export const SAVE_PLAN = 'Save Plan';
export const SAVE_AS_DRAFT = 'Save as a Draft';
export const SAVE_FINALIZED_PLAN = 'Save Finalized Plan';
export const SAVE_ASSIGNMENTS = 'Save Assignments';
export const NO_PLANS_LOADED_MESSAGE = 'No plans found...';
export const SELECT = 'Select';
export const SELECT_USERNAME = 'Select username';

export const PRACTITIONER = 'Practitioner';
export const PRACTITIONERS = 'Practitioners';
export const SAVE = 'Save';
export const CLEAR = 'Clear';
export const ACTIVE = 'Active';
export const ASSIGN = 'Assign';
export const NO_TEAMS_LOADED_MESSAGE = 'No teams loaded...';
export const SELECT_TEAMS_TO_ASSIGN = 'Select Teams to Assign';
export const ASSIGN_TEAMS = 'Assign Teams';
export const TEAMS_ASSIGNED = 'Teams Assigned';
export const YES = 'Yes';
export const NO = 'No';
export const EDIT = 'Edit';
export const ADD = 'Add';
export const REMOVE = 'Remove';

export const IRS_PLANS = 'IRS Plans';
export const COUNTRY = 'Country';
export const NEW_TITLE = 'New';
export const PLAN_TITLE = 'Plan';
export const PLAN_STATUS = 'Plan Status';
export const TEAMS = 'Teams';
export const TEAM = 'Team';
export const EDIT_TEAM = 'Edit Team';
export const NEW_TEAM = 'New Team';
export const TEAMS_ASSIGNMENT = 'Team Assignment';
export const SEARCH = 'Search';
export const ACTION = 'Action';
export const ACTIONS = 'Actions';
export const IDENTIFIER = 'Identifier';
export const USERNAME = 'Username';
export const VIEW = 'view';

export const REACTIVE = 'Reactive';
export const REACTIVE_INVESTIGATION = 'Reactive Investigation';
export const REQUIRED = 'Required';
export const SAVING = 'Saving';
export const PLANS = 'Manage Plans';
export const PLANNING = 'Planning';
export const MONITOR = 'Monitor';
export const ADMIN = 'Admin';
export const DRAFTS_PARENTHESIS = '(drafts)';
export const CONFIRM = 'Confirm';
export const CANCEL = 'Cancel';
export const JURISDICTION = 'Jurisdiction';
export const ADMIN_LEVEL = 'Admin Level';
export const DISTRICT = 'District';
export const CANTON = 'Canton';
export const FI_REASON = 'FI Reason';
export const FI_STATUS = 'FI Status';
export const MEASURE = 'Measure';
export const MARK_AS_COMPLETE = 'Mark as complete';
export const STRUCTURES = 'structure(s)';
export const PERSONS = 'person(s)';

export const FOCUS_AREA_HEADER = 'Focus Area';
export const SPRAY_AREA_HEADER = 'Spray Area';
export const INTERVENTION_TYPE_LABEL = 'Intervention Type';
export const FOCUS_INVESTIGATION_STATUS_LABEL = 'Focus Investigation Status';
export const FOCUS_INVESTIGATION_STATUS_REASON = 'Focus Investigation Reason';
export const CASE_NUMBER = 'Case Number';
export const LAST_MODIFIED = 'Last Modified';
export const TITLE = 'Title';
export const PLAN_TITLE_LABEL = 'Plan Title';
export const PLAN_START_DATE_LABEL = 'Plan Start Date';
export const PLAN_END_DATE_LABEL = 'Plan End Date';
export const ACTIVITIES_LABEL = 'Activities';
export const DESCRIPTION_LABEL = 'Description';
export const QUANTITY_LABEL = 'Quantity';
export const PRIORITY_LABEL = 'Priority';
export const ADD_ACTIVITY = 'Add Activity';
export const REASON_HEADER = 'Reason';
export const STATUS_HEADER = 'Status';
export const CASE_NOTIF_DATE_HEADER = 'Case Notif. Date';
export const CASE_CLASSIFICATION_HEADER = 'Case Class.';
export const CASE_CLASSIFICATION_LABEL = 'Case Classification';
export const DEFINITIONS = 'Definitions';
export const TYPE_LABEL = 'Type';
export const LOCATIONS = 'Locations';

export const PROVINCE = 'Province';
export const HOME = 'Home';
export const FOCUS_INVESTIGATION = 'Focus Investigation';
export const CURRENT_FOCUS_INVESTIGATION = 'Current Focus Investigations';
export const COMPLETE_FOCUS_INVESTIGATION = 'Complete Focus Investigations';
export type COMPLETE_FOCUS_INVESTIGATION = typeof COMPLETE_FOCUS_INVESTIGATION;
export const INVESTIGATION = 'Investigation';
export const INTERVENTION = 'Intervention';
export const FOCUS_AREA_INFO = 'Focus Area Information';
export const FOCUS_INVESTIGATIONS = 'Focus Investigations';
export const ASSIGN_PRACTITIONERS_TO_ORG = 'Assign Practitioners to %s';
export const DATE_IS_REQUIRED = 'Date is Required';
export const NAME_IS_REQUIRED = 'Name is Required';
export const NUMERATOR_OF_DENOMINATOR_UNITS = '%s of %s %s';
export const MARK_PLAN_AS_COMPLETE = 'Mark %s as complete';
export const FIS_IN_JURISDICTION = 'Focus Investigations in %s';
export const FI_IN_JURISDICTION = 'Current Focus Investigations in %s';
export const PROGRESS = 'Progress';
export const ASSIGN_JURISDICTIONS = 'Assign Jurisdictions';
export const SELECT_JURISDICTIONS = 'Select Jurisdictions';
export const ADD_CODED_ACTIVITY = 'Add %s Activity';
export const ORGANIZATION_DETAILS = 'Team Details';
export const TEAM_MEMBERS = 'Team Members';
export const ONE_DAY_TO_GO = '1 day to go';
export const ZERO_DAYS_TO_GO = '0 days to go';
export const PRACTITIONERS_ASSIGNED_TO_ORG = '%s Practitioners successfully assigned to %s';
export const PRACTITIONER_REMOVED_FROM_ORG = '%s-%s successfully removed from %s';
export const ORGANIZATION_EDITED_SUCCESSFULLY = `Team edited successfully.`;
export const ORGANIZATION_CREATED_SUCCESSFULLY = `Team created successfully`;
export const DRAFT_PLAN_TITLE = '%s (draft)';
export const EDIT_PRACTITIONER = 'Edit Practitioner';
export const NEW_PRACTITIONER = 'New Practitioner';
export const SAVE_PRACTITIONER = 'Save Practitioner';

export const ROUTINE_TITLE = 'Routine';
export const ROUTINE_INVESTIGATION_TITLE = 'Routine Investigation';
export const BCC_ACTIVITY = 'Behaviour Change Communication';
export const BCC_ACTIVITY_DESCRIPTION = 'Conduct BCC activity';
export const BCC_GOAL_DESCRIPTION = 'Complete at least 1 BCC activity for the operational area';
export const BCC_GOAL_MEASURE = 'Number of BCC Activities Completed';
export const IRS_ACTIVITY = 'Spray Structures';
export const IRS_ACTIVITY_DESCRIPTION =
  'Visit each structure in the operational area and attempt to spray';
export const IRS_GOAL_DESCRIPTION = 'Spray structures in the operational area';
export const IRS_GOAL_MEASURE = 'Percent of structures sprayed';
export const BEDNET_ACTIVITY = 'Bednet Distribution';
export const BEDNET_ACTIVITY_DESCRIPTION =
  'Visit 100% of residential structures in the operational area and provide nets';
export const BEDNET_GOAL_MEASURE = 'Percent of residential structures received nets';
export const BLOOD_SCREENING_ACTIVITY = 'Blood screening';
export const BLOOD_SCREENING_ACTIVITY_DESCRIPTION =
  'Visit all residential structures (100%) within a 1 km radius of a confirmed index case and test each registered person';
export const BLOOD_SCREENING_GOAL_MEASURE = 'Number of registered people tested';
export const CASE_CONFIRMATION_ACTIVITY = 'Case Confirmation';
export const CASE_CONFIRMATION_ACTIVITY_DESCRIPTION = 'Confirm the index case';
export const CASE_CONFIRMATION_ACTIVITY_GOAL_MEASURE = 'Number of cases confirmed';
export const RACD_ACTIVITY = 'Family Registration';
export const RACD_ACTIVITY_DESCRIPTION =
  'Register all families & family members in all residential structures enumerated (100%) within the operational area';
export const RACD_ACTIVITY_GOAL_MEASURE =
  'Percent of residential structures with full family registration';
export const LARVAL_DIPPING_ACTIVITY = 'Larval Dipping';
export const LARVAL_DIPPING_ACTIVITY_DESCRIPTION =
  'Perform a minimum of three larval dipping activities in the operational area';
export const LARVAL_DIPPING_GOAL_MEASURE = 'Number of larval dipping activities completed';
export const MOSQUITO_COLLECTION_ACTIVITY = 'Mosquito Collection';
export const MOSQUITO_COLLECTION_ACTIVITY_DESCRIPTION =
  'Set a minimum of three mosquito collection traps and complete the mosquito collection process';
export const MOSQUITO_COLLECTION_GOAL_MEASURE =
  'Number of mosquito collection activities completed';

// TODO ? - do the below 2 belong here or in a settings file
export const ORGANIZATION_LABEL = TEAM;
export const ORGANIZATIONS_LABEL = TEAMS;
