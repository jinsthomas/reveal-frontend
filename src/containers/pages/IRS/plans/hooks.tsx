import { SupersetFormData } from '@onaio/superset-connector/dist/types';
import React, { useEffect, useState } from 'react';
import { ActionCreator } from 'redux';
import { SUPERSET_IRS_REPORTING_PLANS_SLICE } from '../../../../configs/env';
import { displayError } from '../../../../helpers/errors';
import supersetFetch from '../../../../services/superset';

interface AsyncSupersetPlans<TData, TAction> {
  superset: typeof supersetFetch;
  data: TData[];
  fetchPlans: ActionCreator<TAction>;
  supersetSlice: string;
  supersetOptions: SupersetFormData | null;
}

export const useAsyncSupersetPlans = <TData, TAction>({
  superset = supersetFetch,
  data = [],
  fetchPlans,
  supersetSlice = SUPERSET_IRS_REPORTING_PLANS_SLICE,
  supersetOptions = null,
}: AsyncSupersetPlans<TData, TAction>) => {
  const [loading, setLoading] = useState<boolean>(false);
  const mounted = React.useRef<boolean>(false);

  /** async function to load the data */
  async function loadData() {
    try {
      setLoading(data.length === 0); // only set loading when there are no plans
      const asyncOperation = supersetOptions
        ? superset(supersetSlice, supersetOptions)
        : superset(supersetSlice);

      await asyncOperation.then((result: TData[]) => fetchPlans(result));
    } catch (e) {
      throw e;
    } finally {
      if (mounted) {
        setLoading(false);
      }
    }
  }

  useEffect(() => {
    mounted.current = true;
    loadData().catch(error => displayError(error));
    return () => {
      mounted.current = false;
    };
  }, []);

  return loading;
};
