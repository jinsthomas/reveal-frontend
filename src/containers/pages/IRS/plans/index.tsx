import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import { SUPERSET_IRS_REPORTING_PLANS_SLICE } from '../../../../configs/env';
import { IRS_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_IRS_PLAN_URL } from '../../../../constants';
import { getQueryParams, RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import IRSPlansReducer, {
  fetchIRSPlans,
  GenericPlan,
  makeIRSPlansArraySelector,
  reducerName as IRSPlansReducerName,
} from '../../../../store/ducks/generic/plans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the IRS plan definitions reducer */
reducerRegistry.register(IRSPlansReducerName, IRSPlansReducer);

/** interface for IRS PlanList props */
interface PlanListProps extends GenericPlanListProps {
  ownProps: any;
}

/** Simple component that loads a privew list of IRS plans */
const IRSPlansList = (props: PlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for IRSPlansList */
const defaultProps: PlanListProps = {
  fetchPlans: fetchIRSPlans,
  ownProps: {},
  pageTitle: IRS_PLANS,
  pageUrl: REPORT_IRS_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_IRS_REPORTING_PLANS_SLICE,
};

IRSPlansList.defaultProps = defaultProps;

export { IRSPlansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: GenericPlan[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const plans = makeIRSPlansArraySelector()(state, { plan_title: searchedTitle });

  return {
    ...ownProps,
    plans,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: fetchIRSPlans };

/** Connected ActiveFI component */
const ConnectedIRSPlansList = connect(mapStateToProps, mapDispatchToProps)(IRSPlansList);

export default ConnectedIRSPlansList;
