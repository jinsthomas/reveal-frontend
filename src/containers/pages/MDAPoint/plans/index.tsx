import reducerRegistry from '@onaio/redux-reducer-registry';
import React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { Store } from 'redux';
import { SUPERSET_MDA_POINT_REPORTING_PLANS_SLICE } from '../../../../configs/env';
import { MDA_POINT_PLANS } from '../../../../configs/lang';
import { QUERY_PARAM_TITLE, REPORT_MDA_POINT_PLAN_URL } from '../../../../constants';
import { getQueryParams } from '../../../../helpers/utils';
import { RouteParams } from '../../../../helpers/utils';
import supersetFetch from '../../../../services/superset';
import MDAPointPlansReducer, {
  fetchMDAPointPlans,
  makeMDAPointPlansArraySelector,
  MDAPointPlan,
  reducerName as MDAPointPlansReducerName,
} from '../../../../store/ducks/generic/MDAPointPlans';
import { GenericPlanListProps, GenericPlansList } from '../../GenericPlansList';

/** register the MDA plan definitions reducer */
reducerRegistry.register(MDAPointPlansReducerName, MDAPointPlansReducer);

// /** interface for MDA PlanList props */
interface PlanListProps extends GenericPlanListProps {
  ownProps: any;
}

/** Simple component that loads a preview list of MDA plans */
const MDAPointPlansList = (props: PlanListProps & RouteComponentProps) => {
  return (
    <div>
      <GenericPlansList {...props} />
    </div>
  );
};

/** Declare default props for MDAPointPlansList */
const defaultProps: PlanListProps = {
  fetchPlans: fetchMDAPointPlans,
  ownProps: {},
  pageTitle: MDA_POINT_PLANS,
  pageUrl: REPORT_MDA_POINT_PLAN_URL,
  plans: [],
  service: supersetFetch,
  supersetReportingSlice: SUPERSET_MDA_POINT_REPORTING_PLANS_SLICE,
};

MDAPointPlansList.defaultProps = defaultProps;

export { MDAPointPlansList };

/** Connect the component to the store */

/** interface to describe props from mapStateToProps */
interface DispatchedStateProps extends RouteComponentProps<RouteParams> {
  plans: MDAPointPlan[];
}

/** map state to props */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps<RouteParams>
): DispatchedStateProps => {
  const searchedTitle = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const MDAPointPlansArray = makeMDAPointPlansArraySelector()(state, { plan_title: searchedTitle });

  return {
    ...ownProps,
    plans: MDAPointPlansArray,
  };
};

/** map dispatch to props */
const mapDispatchToProps = { fetchPlans: fetchMDAPointPlans };

/** Connected ActiveFI component */
const ConnectedMDAPointPlansList = connect(mapStateToProps, mapDispatchToProps)(MDAPointPlansList);

export default ConnectedMDAPointPlansList;
