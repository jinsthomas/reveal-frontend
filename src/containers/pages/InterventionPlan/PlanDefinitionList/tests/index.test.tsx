import { renderTable } from '@onaio/drill-down-table';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import ConnectedPlanDefinitionList, { PlanDefinitionList } from '../';
import { NO_DATA_FOUND } from '../../../../../configs/lang';
import { PLAN_LIST_URL } from '../../../../../constants';
import * as planDataLoaders from '../../../../../helpers/dataLoading/plans';
import store from '../../../../../store';
import { fetchPlanDefinitions } from '../../../../../store/ducks/opensrp/PlanDefinition';
import * as fixtures from '../../../../../store/ducks/opensrp/PlanDefinition/tests/fixtures';
import { fetchPlansByUser } from '../../../../../store/ducks/opensrp/planIdsByUser';

jest.mock('../../../../../configs/env');
/* tslint:disable-next-line no-var-requires */
const fetch = require('jest-fetch-mock');

const history = createBrowserHistory();

jest.mock('../../../../../components/forms/UserFilter', () => {
  return {
    UserSelectFilter: () => <div id="user-filter-mock">Filter users here</div>,
  };
});

describe('components/InterventionPlan/PlanDefinitionList', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  beforeAll(() => {
    global.Date.now = jest.fn(() => new Date('2019-04-07T10:20:30Z').getTime());
  });

  it('renders without crashing', () => {
    const mockList = jest.fn(async () => {
      return Promise.reject(new Error('random error'));
    });
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    const props = {
      history,
      location: {
        hash: '',
        pathname: PLAN_LIST_URL,
        search: '',
        state: undefined,
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}/`,
        url: `${PLAN_LIST_URL}/`,
      },
      plans: fixtures.plans,
      service: mockClass,
    };
    shallow(
      <Router history={history}>
        <PlanDefinitionList {...props} />
      </Router>
    );
  });

  it('renders plan definition list correctly', async () => {
    const mockList = jest.fn(async () => []);
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    fetch.mockResponseOnce(JSON.stringify(fixtures.plansJSON));
    const props = {
      history,
      location: {
        hash: '',
        key: 'key',
        pathname: PLAN_LIST_URL,
        search: '',
        state: undefined,
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}/`,
        url: `${PLAN_LIST_URL}/`,
      },
      plans: fixtures.plans,
      service: mockClass,
    };
    const wrapper = mount(
      <Router history={history}>
        <PlanDefinitionList {...props} />
      </Router>
    );
    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();
    expect(wrapper.find('HeaderBreadcrumb').props()).toMatchSnapshot('bread crumb props');
    expect(toJson(wrapper.find('Row').at(0))).toMatchSnapshot('row heading');
    expect(toJson(wrapper.find('HelmetWrapper'))).toMatchSnapshot('helmet');
    wrapper.unmount();
  });

  it('sort works correctly for intervention type column', async () => {
    const mockList = jest.fn(async () => []);
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    fetch.mockResponseOnce(fixtures.plansJSON);
    const props = {
      history,
      location: {
        hash: '',
        key: 'key',
        pathname: PLAN_LIST_URL,
        search: '',
        state: undefined,
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}/`,
        url: `${PLAN_LIST_URL}/`,
      },
      plans: fixtures.plans,
      service: mockClass,
    };
    const wrapper = mount(
      <Router history={history}>
        <PlanDefinitionList {...props} />
      </Router>
    );
    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();

    renderTable(wrapper, 'before clicking on sort');

    // find column head for intervention type and simulate click.
    expect(
      wrapper
        .find('.table .th')
        .at(1)
        .text()
    ).toMatchInlineSnapshot(`"Intervention Type"`);
    wrapper
      .find('.table .th')
      .at(1)
      .simulate('click');
    wrapper.update();

    renderTable(wrapper, 'after sorting by the intervention type');
    wrapper.unmount();
  });

  it('handles search correctly', async () => {
    const mockList = jest.fn(async () => []);
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    store.dispatch(fetchPlanDefinitions(fixtures.plans));

    const props = {
      fetchPlans: jest.fn(),
      history,
      location: {
        pathname: PLAN_LIST_URL,
        search: '?title=Mosh',
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}`,
        url: `${PLAN_LIST_URL}`,
      },
      service: mockClass,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPlanDefinitionList {...props} />
        </Router>
      </Provider>
    );
    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();
    expect(
      wrapper
        .find('.tbody .tr .td')
        .find('Link')
        .at(0)
        .text()
    ).toEqual('A Test By Mosh');
  });

  it('handles a case insensitive search', async () => {
    const mockList = jest.fn(async () => []);
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    store.dispatch(fetchPlanDefinitions(fixtures.plans));

    const props = {
      fetchPlans: jest.fn(),
      history,
      location: {
        pathname: PLAN_LIST_URL,
        search: '?title=MOSH',
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}`,
        url: `${PLAN_LIST_URL}`,
      },
      service: mockClass,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPlanDefinitionList {...props} />
        </Router>
      </Provider>
    );
    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();

    expect(
      wrapper
        .find('.tbody .tr .td')
        .find('Link')
        .at(0)
        .text()
    ).toEqual('A Test By Mosh');
  });

  it('renders empty table if no search matches', async () => {
    const mockList = jest.fn(async () => []);
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    store.dispatch(fetchPlanDefinitions(fixtures.plans));

    const props = {
      fetchPlans: jest.fn(),
      history,
      location: {
        pathname: PLAN_LIST_URL,
        search: '?title=asdfasdf',
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}`,
        url: `${PLAN_LIST_URL}`,
      },
      service: mockClass,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPlanDefinitionList {...props} />
        </Router>
      </Provider>
    );
    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();
    expect(wrapper.text().includes(NO_DATA_FOUND)).toBeTruthy();
  });

  it('filters correctly when searching by userName', async () => {
    (planDataLoaders as any).loadPlansByUserFilter = async () => {
      return;
    };
    const mockList = jest.fn(async () => []);
    const mockClass: any = jest.fn().mockImplementation(() => {
      return {
        list: mockList,
      };
    });
    store.dispatch(fetchPlanDefinitions(fixtures.plans));
    const userName = 'macTavish';
    store.dispatch(fetchPlansByUser([fixtures.plans[0]], userName));
    const props = {
      fetchPlans: jest.fn(),
      history,
      location: {
        pathname: PLAN_LIST_URL,
        search: `?user=${userName}`,
      },
      match: {
        isExact: true,
        params: {},
        path: `${PLAN_LIST_URL}`,
        url: `${PLAN_LIST_URL}`,
      },
      service: mockClass,
    };
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedPlanDefinitionList {...props} />
        </Router>
      </Provider>
    );
    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();
    // only one plan is rendered fixtures.plan1
    expect(wrapper.find('.tbody .tr').text()).toMatchInlineSnapshot(
      `"A2-Lusaka Akros Test Focus 2FIactive2019-05-19"`
    );
  });
});
