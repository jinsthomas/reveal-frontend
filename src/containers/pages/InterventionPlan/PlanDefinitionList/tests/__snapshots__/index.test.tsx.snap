// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InterventionPlan/PlanDefinitionList renders plan definition list correctly: bread crumb props 1`] = `
Object {
  "currentPage": Object {
    "label": "Manage Plans",
    "url": "/plans/list",
  },
  "pages": Array [
    Object {
      "label": "Home",
      "url": "/",
    },
  ],
}
`;

exports[`components/InterventionPlan/PlanDefinitionList renders plan definition list correctly: helmet 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
>
  <SideEffect(NullComponent)
    defer={true}
    encodeSpecialCharacters={true}
    title="Manage Plans"
    titleAttributes={Object {}}
  >
    <NullComponent
      defer={true}
      encodeSpecialCharacters={true}
      title="Manage Plans"
      titleAttributes={Object {}}
    />
  </SideEffect(NullComponent)>
</HelmetWrapper>
`;

exports[`components/InterventionPlan/PlanDefinitionList renders plan definition list correctly: row heading 1`] = `
<Row
  tag="div"
>
  <div
    className="row"
  >
    <Col
      md={8}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <div
        className="col-md-8"
      >
        <h3
          className="mt-3 mb-3 page-title"
        >
          Manage Plans
        </h3>
      </div>
    </Col>
    <Col
      md={4}
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <div
        className="col-md-4"
      >
        <LinkAsButton
          classNameProp="focus-investigation btn btn-primary float-right mt-3 mb-3"
          text="Add Plan"
          to="/plans/new"
        >
          <Link
            className="focus-investigation btn btn-primary float-right mt-3 mb-3"
            replace={false}
            to="/plans/new"
          >
            <a
              className="focus-investigation btn btn-primary float-right mt-3 mb-3"
              href="/plans/new"
              onClick={[Function]}
            >
              Add Plan
            </a>
          </Link>
        </LinkAsButton>
      </div>
    </Col>
  </div>
</Row>
`;
